substrate/primitives/runtime-interface/src/pass_by.rs

```diff
@@ -15,419 +15,667 @@
+where
+	U: RIType,
+{
+	type FFIType = <U as RIType>::FFIType;
+	type Inner = T;
+}
+
+#[cfg(not(substrate_runtime))]
+impl<T, U> IntoFFIValue for ReturnAs<T, U>
+where
+	U: RIType + IntoFFIValue + Primitive,
+	<U as RIType>::Inner: From<Self::Inner>,
+{
+	fn into_ffi_value(
+		value: Self::Inner,
+		context: &mut dyn FunctionContext,
+	) -> Result<Self::FFIType> {
+		let value: <U as RIType>::Inner = value.into();
+		<U as IntoFFIValue>::into_ffi_value(value, context)
 	}
 }
 
-/// The type is passed as the inner type.
-impl<T: PassByInner<Inner = I>, I: RIType> RIType for Inner<T, I> {
-	type FFIType = I::FFIType;
+#[cfg(substrate_runtime)]
+impl<T, U> FromFFIValue for ReturnAs<T, U>
+where
+	U: RIType + FromFFIValue + Primitive,
+	Self::Inner: TryFrom<U::Inner>,
+{
+	fn from_ffi_value(arg: Self::FFIType) -> Self::Inner {
+		let value = <U as FromFFIValue>::from_ffi_value(arg);
+		match Self::Inner::try_from(value) {
+			Ok(value) => value,
+			Err(_) => {
+				panic!(
+					"failed to convert '{}' (passed as '{}') into a '{}' when marshalling a hostcall's return value through the FFI boundary",
+					type_name::<U::Inner>(),
+					type_name::<Self::FFIType>(),
+					type_name::<Self::Inner>()
+				);
+			},
+		}
+	}
 }
 
-/// The implementation of the pass by enum strategy. This strategy uses an `u8` internally to pass
-/// the enum between wasm and the host. So, this strategy only supports enums with unit variants.
+/// (DEPRECATED) Return `T` as a blob of bytes into the runtime.
 ///
-/// Use this type as associated type for [`PassBy`] to implement this strategy for a type.
+/// Uses `T::AsRef<[u8]>` to cast `T` into a `&[u8]`, allocates runtime memory
+/// using the legacy allocator, copies the slice into the runtime memory, and
+/// returns a pointer to it.
 ///
-/// This type expects the type that wants to implement this strategy as generic parameter. Besides
-/// that the type needs to implement `TryFrom<u8>` and `From<Self> for u8`.
+/// THIS STRATEGY IS DEPRECATED; DO NOT USE FOR NEW HOST FUNCTIONS!
 ///
-/// [`PassByEnum`](derive.PassByEnum.html) is a derive macro to implement this strategy.
-///
-/// # Example
-/// ```
-/// # use sp_runtime_interface::pass_by::{PassBy, Enum};
-/// #[derive(Clone, Copy)]
-/// enum Test {
-///     Test1,
-///     Test2,
-/// }
-///
-/// impl From<Test> for u8 {
-///     fn from(val: Test) -> u8 {
-///         match val {
-///             Test::Test1 => 0,
-///             Test::Test2 => 1,
-///         }
-///     }
-/// }
+/// Raw FFI type: `u32` (a pointer to the byte blob)
+pub struct AllocateAndReturnPointer<T, const N: usize>(PhantomData<(T, [u8; N])>);
+
+impl<T, const N: usize> RIType for AllocateAndReturnPointer<T, N> {
+	type FFIType = u32;
+	type Inner = T;
+}
+
+#[cfg(not(substrate_runtime))]
+impl<T, const N: usize> IntoFFIValue for AllocateAndReturnPointer<T, N>
+where
+	T: AsRef<[u8]>,
+{
+	fn into_ffi_value(value: Self::Inner, context: &mut dyn FunctionContext) -> Result<u32> {
```

Comment:
Nit: Just to be consistent with the other impls. 
```suggestion
	fn into_ffi_value(value: Self::Inner, context: &mut dyn FunctionContext) -> Result<Self::FFIType> {
```